---
title: "Workflow_glioblastoma_multiforme"
author: "Narine Fischer Albiol"
output: html_document
date: "2024-09-18"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed(150)
```

## Introduction

The following workflow uses R language [@rsoftware] to obtain a a Multidimensional Scaling (MDS) comparing the different samples obtained from the Genomic Data Commons Data Portal [@gdcportal]. RNA-seq expression data in samples of patients with glioblastoma multiforme (GM) was used.

## Needed libraries

The libraries used come from both CRAN [@cran] and Bioconductor [@bioconductor]. 
```{r libraries}
library(TCGAbiolinks)
library(edgeR)
library(geneplotter)
library(limma)
library(sva)
library(dplyr)
```

## Data download

As previously stated, the data will be obtained from the Genomic Data Commons data portal [@gdcportal], a repository with data from projects studying cancer. In this case, the data will be downloaded from the project TCGA-GBM, with data specific to glioblastoma multiforme. Therefore, The results published here are in whole or part based upon data generated by The Cancer Genome Atlas managed by the NCI and NHGRI. Information about TCGA can be found at http://cancergenome.nih.gov .The RNA-seq data obtained during this project will be the one selected. For this step, the package `TCGAbiolinks` will be used [@TCGAbiolinks].
```{r data_download}
# Generate query to obtain the RNA-Seq data from the TCGA-GBM project
query_rnaseq <- GDCquery(
  project = "TCGA-GBM",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  access = "open"
)

# Download the RNA-Seq data
GDCdownload(query_rnaseq, method = "client")

# Prepare RNA-Seq data for analysis, creating a SummarizedExperiment object
rnaseq_data <- GDCprepare(query_rnaseq)
```

In the `rnaseq_data` object, there are 60660 genes by 176 samples. Looking at the data of the columns (samples) (`colData(rnaseq_data)`), it can be seen in `definition` that samples pertain to three different groups: Primary solid Tumor (158), Recurrent Solid Tumor (13), and Solid Tissue Normal (5). 

## Quality assessment : sequencing depth

A DGELISt object will be created for further analyses.
```{r quality_assessment_gender}
# Create a DGEList object
dge <- DGEList(counts=assays(rnaseq_data)$unstranded, genes=rowData(rnaseq_data),
                 samples=colData(rnaseq_data))
```

The sequencing depth per sample will be mapped and colored according to the variable `definition`, which stores the classification of each sample (`Primary solid Tumor`, `Recurrent Solid Tumor`, `Solid Tissue Normal`), in order to see if there are differences in sequencing depth between the three types of samples.
```{r quality_assessment_definition}
# Quality assessment: depth sequencing - definition
ord <- order(dge$sample$lib.size)
rnaseq_data$definition<-factor(rnaseq_data$definition) ## We want to see if there is any difference in sequencing depth due to gender, this should be done with any variable that wants to be investigated to ensure the correct formation of the following graph
barplot(dge$sample$lib.size[ord]/1e6, las=1, ylab="Millions of reads",
        xlab="Samples", col=c("brown", "lightpink", "black")[rnaseq_data$definition[ord]])
legend("topleft", c("Primary solid Tumor","Recurrent solid tumor", "Solid Tissue Normal"), fill=c("brown", "lightpink", "black"), inset=0.01)
```

It can be seen that there might be some differences in sequencng depth between the type of samples, as those samples classified as "Solid Tissue Normal" fall more into the high coverage part of the graph, while those classified as "Recurrent Solid Tumor" are more present at the low coverage region. 

## Within-sample Normalization

In order to adjust for the differences in sequencing depth, normalization within samples will be carried out. For this, counts per million (CPM) will be calculated. Moreover, the logarithm with base 2 will be carried out to reduce variability and make the data more suitable for statistical analysis.
```{r normalization}
# Calculate the log2 of the CPM and add it to the SummarizedExperiment object  
assays(rnaseq_data)$logCPM <- cpm(dge, log=TRUE, prior.count=0.25)
```

The distribution of the expression of the `logCPM` values is plotted to identify if there are any samples that deviates from the rest.
```{r QA_sample_distribution}
# Representation of the distribution of logCPM
par(mfrow=c(1, 2), mar=c(4, 5, 4, 4))
multidensity(as.list(as.data.frame(assays(rnaseq_data)$logCPM)),
               xlab="log2 CPM", legend=NULL,
               main="", cex.axis=1.2, cex.lab=1.5, las=1)
boxplot(assays(rnaseq_data)$logCPM, col="gray", xlab="Samples",
               ylab=expression(log[2] * "CPM"), cex.axis=1.2, cex.lab=1.5, las=1)
```

Most of the samples follow the same expression levels, except for three samples that appear to present higher levels of expression in the multidensity graph, they present a higher density of high `logCPM` values. It was decided to remove them to avoid possible biases.
```{r removing_samples}
# Calculate densities for each sample
densities <- apply(assays(rnaseq_data)$logCPM, 2, density)

# To identify the samples, as the three have their second local maximum at higher values in the x-axis than the rest, this characteristic will be used to identify them and remove them.

## Function to get the local maximums of a density
get_peaks <- function(dens) {
  peaks <- which(diff(sign(diff(dens$y))) == -2)  # Identify the local maximums
  list(x = dens$x[peaks], y = dens$y[peaks])
}

## Find the local maximums for each sample
sample_peaks <- lapply(densities, get_peaks)

## Create an object with the x-axis coordinates of the second local maximums of each sample
second_peak_locations <- sapply(sample_peaks, function(p) p$x[2])

## Find the three samples with their second local maximum having higher values in their coordinates
second_peak_sorted <- sort(second_peak_locations, decreasing = TRUE)
top_three_outliers <- names(second_peak_sorted)[1:3]
cat("Outliers identified for removal:", top_three_outliers, "\n")

# To corroborate that the samples detected are the correct ones, plot densities for all samples, with the top three outliers highlighted in red
plot(density(assays(rnaseq_data)$logCPM[, top_three_outliers[1]]), col = "red", main = "Density Plot of Samples", xlab = "Expression Level", ylim = range(sapply(densities, function(d) d$y)))
lines(density(assays(rnaseq_data)$logCPM[, top_three_outliers[2]]), col = "red")
lines(density(assays(rnaseq_data)$logCPM[, top_three_outliers[3]]), col = "red")
for (i in 1:ncol(assays(rnaseq_data)$logCPM)) {
  if (!colnames(assays(rnaseq_data)$logCPM)[i] %in% top_three_outliers) {
    lines(density(assays(rnaseq_data)$logCPM[, i]), col = "grey")
  }
}
legend("topright", legend = c("Outliers", "Other Samples"), col = c("red", "grey"), lwd = 2)

# Remove outliers from the SummarizedExperiment and the DGEList objects
rnaseq_data <- rnaseq_data[, !colnames(rnaseq_data) %in% top_three_outliers]
dge <- dge[, !colnames(dge) %in% top_three_outliers]
```
hanks to the plot, it is confirmed that the three samples detected are the correct ones.

The multidensity plot and the boxplot are remade to ensure that the removal went accordingly.
```{r QA_sample_distribution_II}
# Representation of the distribution of logCPM
par(mfrow=c(1, 2), mar=c(4, 5, 4, 4))
multidensity(as.list(as.data.frame(assays(rnaseq_data)$logCPM)),
               xlab="log2 CPM", legend=NULL,
               main="", cex.axis=1.2, cex.lab=1.5, las=1)
boxplot(assays(rnaseq_data)$logCPM, col="gray", xlab="Samples",
               ylab=expression(log[2] * "CPM"), cex.axis=1.2, cex.lab=1.5, las=1)
```
This time it can be seen that the distribution is more homogeneous, the samples have been correctly removed. 

## Filtering of lowly-expressed genes

The next step will be filtering out the lowly-expressed genes, as they can introduce artifacts in the analysis. First, a representation of the gene expression is done.
```{r QA_gene_distribution}
# Calculate the average level of expression of each gene
avgexp <- rowMeans(assays(rnaseq_data)$logCPM)

# Plot the average levels of expression 
hist(avgexp, xlab=expression(log[2] * "CPM"), main="", las=1, col="gray")
abline(v=0, col="red", lwd=2)
```

It can be seen that most of the genes are lowly expressed. However, there is a small peak at 5 log2CPM that indicate highly expressed genes. To continue with the analysis, the lowly-expressed genes will be removed. In this case, it has been decided to use a cut-off of 1 log2CPM to remove those genes with lower expression.
```{r LE_genes_removal}
# Create mask using 1 log2CPM as cutoff
mask <- rowMeans(assays(rnaseq_data)$logCPM) > 1

# Create new objects with the filtered genes 
rnaseq_data.filt <- rnaseq_data[mask, ]
dge.filt <- dge[mask, ]
dim(rnaseq_data.filt)
```

A total of 14372 genes that are highly expressed are obtained.

Plotting again the gene expression plot, coloring the 14372 genes in red, the correct selection is confirmed
```{r filtered_genes_representation}
par(mar=c(4, 5, 1, 1))
h <- hist(avgexp, xlab=expression("Expression level (" * log[2] * "CPM)"),
          main="", las=1, col="grey", cex.axis=1.2, cex.lab=1.5)
x <- cut(rowMeans(assays(rnaseq_data.filt)$logCPM), breaks=h$breaks)
lines(h$mids, table(x), type="h", lwd=10, lend=1, col="red")
legend("topright", c("All genes", "Filtered genes"), fill=c("grey", "red"))
```

# Normalization II: between-sample

The next step will be normalizing between samples. The TMM method will be used.
```{r normalization_bet_sample}
dge.filt <- calcNormFactors(dge.filt, method = 'TMM')
```

The raw data is substituted by the normalized data
```{r raw_to_normalized}
# Replace the raw log2 CPM units in the corresponding assay element of the `SummarizedExperiment` object, by the normalized ones.
assays(rnaseq_data.filt)$logCPM <- cpm(dge.filt, log=TRUE,
                              normalized.lib.sizes=TRUE)
```

A boxplot is created to see if there are any significant differences in expression between samples to see if there is any outlier between them
```{r repeat_boxplot}
boxplot(assays(rnaseq_data.filt)$logCPM, col="gray", xlab="Samples",
        ylab=expression(log[2] * "CPM"), las=1)
```

No major difference between the samples are observed.

## Batch identification and experimental design

Finally, the Multidimensional Scaling (MDS) plot comparing all smaples is obtained.
```{r batch_identificatio_nmps_graph}
outcome <- rnaseq_data.filt$shortLetterCode
names(outcome) <- colnames(rnaseq_data.filt)
# Set color palette with three distinct colors
color_palette <- c("red", "blue", "green")

# Map shortLetterCode values to colors
unique_codes <- unique(rnaseq_data.filt$shortLetterCode)
color_map <- setNames(color_palette[1:length(unique_codes)], unique_codes)

# Apply the color mapping to the outcome
outcome_colors <- color_map[outcome]

# Plot the MDS with mapped colors
plotMDS(dge.filt, labels=outcome, col=outcome_colors)
legend("topright", legend=unique_codes, col=color_palette, pch=16)
```
It can be seen that the Healthy samples (`NT`) are mainly forming it's own cluster, only one healthy samples is similar to one tumoral sample, while for the rest, a healthy sample chunk and a tumoral chunk can be observed. This signals differences in expression between healthy samples and tumoral samples. However, Primary tumor samples and recurrent tumor samples are not as differentiated, forming a single chunk.